File: src/main/java/cn/yiiguxing/plugin/translate/action/TranslateAndReplaceAction.java
package cn.yiiguxing.plugin.translate.action;

import cn.yiiguxing.plugin.translate.Settings;
import cn.yiiguxing.plugin.translate.Translator;
import cn.yiiguxing.plugin.translate.Utils;
import cn.yiiguxing.plugin.translate.compat.SelectWordUtilCompat;
import cn.yiiguxing.plugin.translate.model.BasicExplain;
import cn.yiiguxing.plugin.translate.model.QueryResult;
import com.intellij.codeInsight.highlighting.HighlightManager;
import com.intellij.codeInsight.lookup.*;
import com.intellij.openapi.actionSystem.AnActionEvent;
import com.intellij.openapi.actionSystem.PlatformDataKeys;
import com.intellij.openapi.application.ApplicationManager;
import com.intellij.openapi.editor.Editor;
import com.intellij.openapi.editor.ScrollType;
import com.intellij.openapi.editor.SelectionModel;
import com.intellij.openapi.editor.markup.EffectType;
import com.intellij.openapi.editor.markup.RangeHighlighter;
import com.intellij.openapi.editor.markup.TextAttributes;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.util.TextRange;
import com.intellij.openapi.util.text.StringUtil;
import com.intellij.openapi.vfs.ReadonlyStatusHandler;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.ui.JBColor;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.*;

/**
 * 翻译并替换
 */
public class TranslateAndReplaceAction extends AutoSelectAction {

    private static final String PATTERN_FIX = "^(\\[[\\u4E00-\\u9FBF]+])+ ";

    private static final TextAttributes HIGHLIGHT_ATTRIBUTES;

    static {
        TextAttributes attributes = new TextAttributes();
        attributes.setEffectType(EffectType.BOXED);
        attributes.setEffectColor(new JBColor(0xFFFF0000, 0xFFFF0000));

        HIGHLIGHT_ATTRIBUTES = attributes;
    }

    private final Settings mSettings;

    public TranslateAndReplaceAction() {
        super(SelectWordUtilCompat.HANZI_CONDITION, true);
        setEnabledInModalContext(false);
        mSettings = Settings.getInstance();
    }

    @NotNull
    @Override
    protected AutoSelectionMode getAutoSelectionMode() {
        return mSettings.getAutoSelectionMode();
    }

    @Override
    protected void onUpdate(AnActionEvent e, boolean active) {
        if (active) {
            final Editor editor = getEditor(e);
            if (editor != null) {
                SelectionModel selectionModel = editor.getSelectionModel();
                if (selectionModel.hasSelection()) {
                    final String selectedText = selectionModel.getSelectedText();
                    if (!Utils.isEmptyOrBlankString(selectedText)) {
                        boolean hasHanzi = false;
                        for (int i = 0; i < selectedText.length(); i++) {
                            if (SelectWordUtilCompat.HANZI_CONDITION.value(selectedText.charAt(i))) {
                                hasHanzi = true;
                                break;
                            }
                        }

                        active = hasHanzi;
                    }
                }
            }
        }
        e.getPresentation().setEnabledAndVisible(active);
    }

    @Override
    protected void onActionPerformed(final AnActionEvent e,
                                     @NotNull final Editor editor,
                                     @NotNull final TextRange selectionRange) {
        VirtualFile virtualFile = e.getData(PlatformDataKeys.VIRTUAL_FILE);
        final Project project = e.getProject();
        if (project == null || (virtualFile != null &&
                ReadonlyStatusHandler.getInstance(project).ensureFilesWritable(virtualFile).hasReadonlyFiles())) {
            return;
        }

        final String text = editor.getDocument().getText(selectionRange);
        if (Utils.isEmptyOrBlankString(text))
            return;

        Translator.getInstance().query(text, new Translator.Callback() {
            @Override
            public void onQuery(@Nullable String query, @Nullable final QueryResult result) {
                if (result == null || !result.isSuccessful())
                    return;

                final List<LookupElement> replaceLookup = getReplaceLookupElements(result);
                if (replaceLookup.isEmpty())
                    return;

                ApplicationManager.getApplication().invokeLater(new Runnable() {
                    @Override
                    public void run() {
                        doReplace(editor, selectionRange, text, replaceLookup);
                    }
                });
            }
        });
    }

    private static void doReplace(@NotNull final Editor editor,
                                  @NotNull final TextRange selectionRange,
                                  @NotNull final String targetText,
                                  @NotNull final List<LookupElement> replaceLookup) {
        if (editor.isDisposed() || editor.getProject() == null ||
                !targetText.equals(editor.getDocument().getText(selectionRange)) ||
                !selectionRange.containsOffset(editor.getCaretModel().getOffset())) {
            return;
        }

        final SelectionModel selectionModel = editor.getSelectionModel();
        final int startOffset = selectionRange.getStartOffset();
        final int endOffset = selectionRange.getEndOffset();
        if (selectionModel.hasSelection()) {
            if (selectionModel.getSelectionStart() != startOffset || selectionModel.getSelectionEnd() != endOffset) {
                return;
            }
        } else {
            selectionModel.setSelection(startOffset, endOffset);
        }

        editor.getScrollingModel().scrollToCaret(ScrollType.MAKE_VISIBLE);
        editor.getCaretModel().moveToOffset(endOffset);

        LookupElement[] items = replaceLookup.toArray(new LookupElement[replaceLookup.size()]);
        final LookupEx lookup = LookupManager.getInstance(editor.getProject()).showLookup(editor, items);

        if (lookup == null) {
            return;
        }

        final HighlightManager highlightManager = HighlightManager.getInstance(editor.getProject());
        final List<RangeHighlighter> highlighters = addHighlight(highlightManager, editor, selectionRange);

        lookup.addLookupListener(new LookupAdapter() {
            @Override
            public void itemSelected(LookupEvent event) {
                disposeHighlight(highlighters);
            }

            @Override
            public void lookupCanceled(LookupEvent event) {
                selectionModel.removeSelection();
                disposeHighlight(highlighters);
            }
        });
    }

    @NotNull
    private static List<RangeHighlighter> addHighlight(@NotNull HighlightManager highlightManager,
                                                       @NotNull Editor editor,
                                                       @NotNull TextRange selectionRange) {
        final ArrayList<RangeHighlighter> highlighters = new ArrayList<RangeHighlighter>();
        highlightManager.addOccurrenceHighlight(editor, selectionRange.getStartOffset(), selectionRange.getEndOffset(),
                HIGHLIGHT_ATTRIBUTES, 0, highlighters, null);

        for (RangeHighlighter highlighter : highlighters) {
            highlighter.setGreedyToLeft(true);
            highlighter.setGreedyToRight(true);
        }

        return highlighters;
    }

    private static void disposeHighlight(@NotNull List<RangeHighlighter> highlighters) {
        for (RangeHighlighter highlighter : highlighters) {
            highlighter.dispose();
        }
    }

    @NotNull
    private static List<LookupElement> getReplaceLookupElements(@NotNull QueryResult result) {
        final List<LookupElement> replaceLookup;

        BasicExplain basicExplain = result.getBasicExplain();
        if (basicExplain != null) {
            replaceLookup = getReplaceLookupElements(Utils.expandExplain(basicExplain.getExplains()));
        } else {
            replaceLookup = getReplaceLookupElements(result.getTranslation());
        }

        return replaceLookup;
    }

    @NotNull
    private static List<LookupElement> getReplaceLookupElements(@Nullable String[] explains) {
        if (explains == null || explains.length == 0)
            return Collections.emptyList();

        final Set<LookupElement> camel = new LinkedHashSet<LookupElement>();
        final Set<LookupElement> pascal = new LinkedHashSet<LookupElement>();
        final Set<LookupElement> lowerWithUnder = new LinkedHashSet<LookupElement>();
        final Set<LookupElement> capsWithUnder = new LinkedHashSet<LookupElement>();
        final Set<LookupElement> withSpace = new LinkedHashSet<LookupElement>();

        final StringBuilder camelBuilder = new StringBuilder();
        final StringBuilder pascalBuilder = new StringBuilder();
        final StringBuilder lowerWithUnderBuilder = new StringBuilder();
        final StringBuilder capsWithUnderBuilder = new StringBuilder();
        final StringBuilder withSpaceBuilder = new StringBuilder();

        for (String explain : explains) {
            List<String> words = fixAndSplitForVariable(explain);
            if (words == null || words.isEmpty()) {
                continue;
            }

            camelBuilder.setLength(0);
            pascalBuilder.setLength(0);
            lowerWithUnderBuilder.setLength(0);
            capsWithUnderBuilder.setLength(0);
            withSpaceBuilder.setLength(0);

            build(words, camelBuilder, pascalBuilder, lowerWithUnderBuilder, capsWithUnderBuilder, withSpaceBuilder);

            camel.add(LookupElementBuilder.create(camelBuilder.toString()));
            pascal.add(LookupElementBuilder.create(pascalBuilder.toString()));
            lowerWithUnder.add(LookupElementBuilder.create(lowerWithUnderBuilder.toString()));
            capsWithUnder.add(LookupElementBuilder.create(capsWithUnderBuilder.toString()));
            withSpace.add(LookupElementBuilder.create(withSpaceBuilder.toString()));
        }

        final Set<LookupElement> result = new LinkedHashSet<LookupElement>();
        result.addAll(camel);
        result.addAll(pascal);
        result.addAll(lowerWithUnder);
        result.addAll(capsWithUnder);
        result.addAll(withSpace);

        return Collections.unmodifiableList(new ArrayList<LookupElement>(result));
    }

    private static void build(@NotNull final List<String> words,
                              @NotNull final StringBuilder camel,
                              @NotNull final StringBuilder pascal,
                              @NotNull final StringBuilder lowerWithUnder,
                              @NotNull final StringBuilder capsWithUnder,
                              @NotNull final StringBuilder withSpace) {
        for (int i = 0; i < words.size(); i++) {
            String word = words.get(i);

            if (i > 0) {
                lowerWithUnder.append('_');
                capsWithUnder.append('_');
                withSpace.append(' ');
            }

            withSpace.append(word);

            if (i == 0) {
                word = sanitizeJavaIdentifierStart(word);
            }

            String capitalized = StringUtil.capitalizeWithJavaBeanConvention(word);
            String lowerCase = word.toLowerCase();

            camel.append(i == 0 ? lowerCase : capitalized);
            pascal.append(capitalized);
            lowerWithUnder.append(lowerCase);
            capsWithUnder.append(word.toUpperCase());
        }
    }

    @Nullable
    private static List<String> fixAndSplitForVariable(@NotNull String explains) {
        String explain = Utils.splitExplain(explains)[1];
        if (Utils.isEmptyOrBlankString(explain)) {
            return null;
        }

        String fixed = explain.replaceFirst(PATTERN_FIX, "");
        return StringUtil.getWordsIn(fixed);
    }

    private static String sanitizeJavaIdentifierStart(@NotNull String name) {
        return Character.isJavaIdentifierStart(name.charAt(0)) ? name : "_" + name;
    }

}

File: src/main/kotlin/cn/yiiguxing/plugin/translate/action/TranslateAndReplaceAction.kt
package cn.yiiguxing.plugin.translate.action

import cn.yiiguxing.plugin.translate.adaptedMessage
import cn.yiiguxing.plugin.translate.message
import cn.yiiguxing.plugin.translate.trans.Lang
import cn.yiiguxing.plugin.translate.trans.TranslateListener
import cn.yiiguxing.plugin.translate.trans.Translation
import cn.yiiguxing.plugin.translate.trans.TranslationNotifications
import cn.yiiguxing.plugin.translate.ui.SpeedSearchListPopupStep
import cn.yiiguxing.plugin.translate.ui.showListPopup
import cn.yiiguxing.plugin.translate.util.*
import com.intellij.codeInsight.lookup.*
import com.intellij.lang.Language
import com.intellij.openapi.actionSystem.*
import com.intellij.openapi.application.ModalityState
import com.intellij.openapi.application.WriteAction
import com.intellij.openapi.command.CommandProcessor
import com.intellij.openapi.editor.Document
import com.intellij.openapi.editor.Editor
import com.intellij.openapi.editor.ScrollType
import com.intellij.openapi.editor.markup.*
import com.intellij.openapi.editor.textarea.TextComponentEditor
import com.intellij.openapi.editor.textarea.TextComponentEditorImpl
import com.intellij.openapi.fileTypes.PlainTextLanguage
import com.intellij.openapi.progress.impl.BackgroundableProcessIndicator
import com.intellij.openapi.progress.util.AbstractProgressIndicatorExBase
import com.intellij.openapi.project.Project
import com.intellij.openapi.ui.popup.PopupStep
import com.intellij.openapi.util.TextRange
import com.intellij.openapi.util.text.StringUtil
import com.intellij.openapi.vfs.ReadonlyStatusHandler
import com.intellij.ui.JBColor
import com.intellij.util.concurrency.EdtScheduledExecutorService
import java.lang.ref.WeakReference
import java.util.*
import java.util.concurrent.TimeUnit
import javax.swing.text.JTextComponent

/**
 * 翻译并替换
 */
class TranslateAndReplaceAction : AutoSelectAction(true, NON_WHITESPACE_CONDITION), PopupAction {

    init {
        isEnabledInModalContext = true
        templatePresentation.text = adaptedMessage("action.TranslateAndReplaceAction.text")
        templatePresentation.description = message("action.TranslateAndReplaceAction.description")
    }

    override val selectionMode: SelectionMode
        get() = Settings.autoSelectionMode

    override val AnActionEvent.editor: Editor?
        get() {
            val dataContext = dataContext
            val editor = CommonDataKeys.EDITOR.getData(dataContext)
            return if (editor != null) {
                editor
            } else {
                val data = PlatformDataKeys.CONTEXT_COMPONENT.getData(dataContext)
                if (data is JTextComponent) {
                    val project = CommonDataKeys.PROJECT.getData(dataContext)
                    TextComponentEditorImpl(project, data)
                } else null
            }
        }

    override fun onUpdate(e: AnActionEvent): Boolean {
        val editor = e.editor?.takeIf { it.document.isWritable } ?: return false
        val selectionModel = editor.selectionModel
        if (selectionModel.hasSelection()) {
            return showReplacementActionInContextMenu(e) &&
                    selectionModel.selectedText?.any(JAVA_IDENTIFIER_PART_CONDITION) ?: false
        }
        return mayTranslateWithNoSelection(e) && showReplacementActionInContextMenu(e) && super.onUpdate(e)
    }

    override fun onActionPerformed(event: AnActionEvent, editor: Editor, selectionRange: TextRange) {
        val language = event.getData(LangDataKeys.LANGUAGE)
        val project = editor.project ?: CommonDataKeys.PROJECT.getData(event.dataContext)
        val document = editor.document
        val isWritable = project
            ?.let { ReadonlyStatusHandler.ensureDocumentWritable(it, document) }
            ?: document.isWritable
        if (!isWritable) {
            return
        }
        if (!TranslateService.translator.checkConfiguration()) {
            return
        }

        // 不要在这下面使用`event`，否则将可能会出现 `cannot share data context between Swing events` 错误。
        // 因为`TranslateService.translator.checkConfiguration()`可能会占用很长的时间并等待，
        // 之后`event`中的`DataContext`就可能变得不可使用了。详见：
        // https://intellij-support.jetbrains.com/hc/en-us/community/posts/207379815-AnActionEvent-cannot-share-data-context-between-Swing-events-

        val editorRef = WeakReference(editor)
        val text = editor.document.getText(selectionRange)
            .takeIf { it.isNotBlank() && it.any(JAVA_IDENTIFIER_PART_CONDITION) }
            ?: return
        val processedText = text.trim().let { if (!it.contains(WHITESPACE)) it.splitWords() else it }
        val primaryLanguage = TranslateService.translator.primaryLanguage

        val indicator = Indicator(project, editorRef).apply { setProgressText(processedText) }
        val modalityState = ModalityState.current()
        fun translate(targetLang: Lang, reTranslate: Boolean = false) {
            if (indicator.checkProcessCanceledAndEditorDisposed()) {
                return
            }
            TranslateService.translate(processedText, Lang.AUTO, targetLang, object : TranslateListener {
                override fun onSuccess(translation: Translation) {
                    if (indicator.checkProcessCanceledAndEditorDisposed()) {
                        return
                    }
                    if (reTranslate
                        && translation.srcLang == Lang.ENGLISH
                        && primaryLanguage != Lang.ENGLISH
                        && targetLang == Lang.ENGLISH
                    ) {
                        val delay = TranslateService.translator.intervalLimit
                        if (delay <= 0) {
                            translate(primaryLanguage)
                        } else {
                            EdtScheduledExecutorService.getInstance()
                                .schedule({ translate(primaryLanguage) }, delay.toLong(), TimeUnit.MILLISECONDS)
                        }
                    } else {
                        indicator.processFinish()
                        val elementsToReplace = createReplaceElements(translation, language)
                        editorRef.get()?.showResultsIfNeeds(selectionRange, text, elementsToReplace)
                    }
                }

                override fun onError(throwable: Throwable) {
                    if (indicator.checkProcessCanceledAndEditorDisposed()) {
                        return
                    }

                    indicator.processFinish()
                    editorRef.get()?.showResultsIfNeeds(selectionRange, text, emptyList())
                    TranslationNotifications.showTranslationErrorNotification(
                        project, message("translate.and.replace.notification.title"), null, throwable
                    )
                }
            }, modalityState)
        }

        if (Settings.selectTargetLanguageBeforeReplacement) {
            editor.showTargetLanguagesPopup { translate(it) }
        } else {
            val targetLang = if (processedText.any(NON_LATIN_CONDITION)) Lang.ENGLISH else primaryLanguage
            translate(targetLang, true)
        }
    }


    private class Indicator(
        val project: Project?,
        val editorRef: WeakReference<Editor>
    ) : BackgroundableProcessIndicator(
        project,
        message("action.TranslateAndReplaceAction.description"),
        null,
        null,
        true
    ) {

        init {
            initDelegate()
            start()
            isIndeterminate = true
            text = message("action.TranslateAndReplaceAction.task.text")
        }

        private fun initDelegate() {
            addStateDelegate(object : AbstractProgressIndicatorExBase() {
                override fun cancel() {
                    // 在用户取消的时候使`progressIndicator`立即结束并且不再显示
                    this@Indicator.processFinish()
                }
            })
        }

        fun setProgressText(text: String) {
            text2 = message("action.TranslateAndReplaceAction.task.text2", text)
        }

        fun checkProcessCanceledAndEditorDisposed(): Boolean {
            if (isCanceled) {
                // no need to finish the progress indicator,
                // because it's already finished in the delegate.
                return true
            }
            if ((project != null && project.isDisposed) || editorRef.get().let { it == null || it.isDisposed }) {
                processFinish()
                return true
            }
            return false
        }
    }


    private class TranslationItemRenderer : LookupElementRenderer<LookupElement>() {
        override fun renderElement(element: LookupElement, presentation: LookupElementPresentation) {
            presentation.itemText = element.lookupString.replace(CRLF, "↩")
        }
    }


    private companion object {

        /** 谷歌翻译的空格符：`  -   　` */
        val SPACES = Regex("[\u00a0\u2000-\u200a\u202f\u205f\u3000]")

        val CRLF = Regex("\r\n|\r|\n")

        val WHITESPACE = Regex("\\s+")

        val HIGHLIGHT_ATTRIBUTES = TextAttributes().apply {
            effectType = EffectType.BOXED
            effectColor = JBColor(0xFF0000, 0xFF0000)
        }

        fun String.fixWhitespace() = replace(SPACES, " ")

        fun Editor.showTargetLanguagesPopup(onChosen: (Lang) -> Unit) {
            val states = TranslationStates
            val languages = TranslateService.translator.supportedTargetLanguages.sortedByDescending {
                if (it == Lang.AUTO) Int.MAX_VALUE else states.getLanguageScore(it)
            }
            val index = languages.indexOf(states.lastReplacementTargetLanguage)

            val step = object : SpeedSearchListPopupStep<Lang>(languages, title = message("title.targetLanguage")) {
                override fun getTextFor(value: Lang): String = value.langName
                override fun onChosen(selectedValue: Lang, finalChoice: Boolean): PopupStep<*>? {
                    onChosen(selectedValue)
                    states.accumulateLanguageScore(selectedValue)
                    states.lastReplacementTargetLanguage = selectedValue
                    return super.onChosen(selectedValue, true)
                }
            }
            if (index >= 0) {
                step.defaultOptionIndex = index
            }

            showListPopup(step, 10)
        }

        fun Editor.canShowPopup(selectionRange: TextRange, targetText: String): Boolean {
            return selectionRange.endOffset <= document.textLength &&
                    targetText == document.getText(selectionRange) &&
                    selectionRange.containsOffset(caretModel.offset)
        }

        fun Editor.tryReplace(selectionRange: TextRange, elementsToReplace: List<String>): Boolean {
            return if (elementsToReplace.size == 1 && Settings.autoReplace) {
                replaceText(selectionRange, elementsToReplace.first())
                true
            } else false
        }

        fun Editor.checkSelection(selectionRange: TextRange): Boolean {
            val startOffset = selectionRange.startOffset
            val endOffset = selectionRange.endOffset
            with(selectionModel) {
                if (hasSelection()) {
                    if (selectionStart != startOffset || selectionEnd != endOffset) {
                        return false
                    }
                } else {
                    setSelection(startOffset, endOffset)
                }
            }

            scrollingModel.scrollToCaret(ScrollType.MAKE_VISIBLE)
            caretModel.moveToOffset(endOffset)

            return true
        }

        fun Editor.beforeShowPopup(
            selectionRange: TextRange,
            targetText: String,
            elementsToReplace: List<String>
        ): Boolean {
            if (isDisposed) {
                return false
            }
            if (!canShowPopup(selectionRange, targetText)) {
                return false
            }
            if (tryReplace(selectionRange, elementsToReplace)) {
                return false
            }
            return checkSelection(selectionRange)
        }

        fun Editor.showResultsIfNeeds(
            selectionRange: TextRange,
            targetText: String,
            elements: List<String>
        ) {
            if (!beforeShowPopup(selectionRange, targetText, elements)) {
                return
            }

            val project = project
            if (this is TextComponentEditor || project == null) {
                showListPopup(selectionRange, elements)
            } else if (!project.isDisposed) {
                showLookup(project, selectionRange, elements)
            }
        }

        fun Editor.showLookup(project: Project, selectionRange: TextRange, elementsToReplace: List<String>) {
            val markupModel = markupModel
            val renderer = TranslationItemRenderer()
            val lookupElements = elementsToReplace.map {
                LookupElementBuilder.create(it).withRenderer(renderer)
            }.toTypedArray()
            val lookup = LookupManager.getInstance(project).showLookup(this, *lookupElements) ?: return
            val highlighter = markupModel.addHighlight(selectionRange)

            lookup.addLookupListener(object : LookupListener {
                override fun itemSelected(event: LookupEvent) {
                    markupModel.removeHighlighter(highlighter)
                }

                override fun lookupCanceled(event: LookupEvent) {
                    selectionModel.removeSelection()
                    markupModel.removeHighlighter(highlighter)
                }
            })
        }

        fun Editor.replaceText(range: TextRange, text: String) {
            CommandProcessor.getInstance().executeCommand(
                project,
                {
                    WriteAction.run<Throwable> {
                        document.startGuardedBlockChecking()
                        try {
                            val offset = document.replaceString(range, text)
                            selectionModel.removeSelection()
                            caretModel.moveToOffset(offset)
                            scrollingModel.scrollToCaret(ScrollType.MAKE_VISIBLE)
                        } finally {
                            document.stopGuardedBlockChecking()
                        }
                    }
                },
                null,
                null
            )
        }

        fun Document.replaceString(range: TextRange, text: String): Int {
            val length = textLength
            val startOffset = minOf(range.startOffset, length)
            val endOffset = minOf(range.endOffset, length)
            replaceString(startOffset, endOffset, text)
            return startOffset + text.length
        }

        fun MarkupModel.addHighlight(selectionRange: TextRange): RangeHighlighter {
            return addRangeHighlighter(
                selectionRange.startOffset,
                selectionRange.endOffset,
                HighlighterLayer.SELECTION - 1,
                HIGHLIGHT_ATTRIBUTES,
                HighlighterTargetArea.EXACT_RANGE
            ).apply {
                isGreedyToLeft = true
                isGreedyToRight = true
            }
        }

        fun Editor.showListPopup(selectionRange: TextRange, elements: List<String>) {
            val step = object : SpeedSearchListPopupStep<String>(elements) {
                override fun onChosen(selectedValue: String, finalChoice: Boolean): PopupStep<*>? {
                    replaceText(selectionRange, selectedValue)
                    return super.onChosen(selectedValue, true)
                }
            }
            showListPopup(step, 10)
        }

        fun createReplaceElements(translation: Translation, language: Language?): List<String> {
            val translationSet = translation.dictDocument?.translations?.toMutableSet() ?: mutableSetOf()
            translation.translation?.let { translationSet.add(it) }
            val items = translationSet.asSequence()
                .filter { it.isNotEmpty() }
                .map { it.fixWhitespace() }
                .toList()

            if (items.isEmpty()) {
                return emptyList()
            }

            if (translation.targetLang != Lang.ENGLISH || language == PlainTextLanguage.INSTANCE) {
                return items.filter { it.isNotBlank() }
            }

            val camel = LinkedHashSet<String>()
            val pascal = LinkedHashSet<String>()
            val original = LinkedHashSet<String>()

            val camelBuilder = StringBuilder()
            val pascalBuilder = StringBuilder()

            val lowerWithSeparatorBuilders = Settings.separators.map { it to StringBuilder() }
            val withSeparator = Settings.separators.map { it to LinkedHashSet<String>() }.toMap()

            for (item in items) {
                original.add(item)
                if (item.length > 50) {
                    continue
                }

                val words: List<String> = StringUtil.getWordsIn(item)
                if (words.isEmpty()) {
                    continue
                }

                camelBuilder.setLength(0)
                pascalBuilder.setLength(0)
                lowerWithSeparatorBuilders.forEach { it.second.setLength(0) }

                build(words, camelBuilder, pascalBuilder, lowerWithSeparatorBuilders)

                camel.add(camelBuilder.toString())
                pascal.add(pascalBuilder.toString())
                for ((separator, builder) in lowerWithSeparatorBuilders) {
                    withSeparator.getValue(separator).apply {
                        add(builder.toString())
                        add(builder.toString().uppercase())
                    }
                }
            }

            return LinkedHashSet<String>()
                .apply {
                    addAll(camel)
                    addAll(pascal)
                    for ((_, elements) in withSeparator) {
                        addAll(elements)
                    }
                    addAll(original)
                }
                .filter { it.isNotBlank() }
        }

        fun build(
            words: List<String>,
            camel: StringBuilder,
            pascal: StringBuilder,
            lowerWithSeparator: List<Pair<Char, StringBuilder>>
        ) {
            for (i in words.indices) {
                val word = if (i == 0) words[i].sanitizeJavaIdentifierStart() else words[i]
                val lowerCase = word.lowercase(Locale.getDefault())
                for ((separator, builder) in lowerWithSeparator) {
                    if (i > 0) {
                        builder.append(separator)
                    }
                    builder.append(lowerCase)
                }

                val capitalized = StringUtil.capitalizeWithJavaBeanConvention(word)
                camel.append(if (i == 0) lowerCase else capitalized)
                pascal.append(capitalized)
            }
        }

        fun String.sanitizeJavaIdentifierStart(): String {
            return if (Character.isJavaIdentifierStart(this[0])) this else "_$this"
        }
    }
}
